zebra-network/src/peer_set/initialize.rs

```diff
@@ -228,36 +214,71 @@ where
     Ok(())
 }
 
-/// Listens for peer connections on `addr`, then sets up each connection as a
-/// Zcash peer.
+/// Open a peer connection listener on `config.listen_addr`,
+/// returning the opened `TcpListener`, and the address it is bound to.
 ///
-/// Uses `handshaker` to perform a Zcash network protocol handshake, and sends
-/// the `Client` result over `tx`.
-#[instrument(skip(tx, handshaker))]
-async fn listen<S>(
-    addr: SocketAddr,
-    mut handshaker: S,
-    tx: mpsc::Sender<PeerChange>,
-) -> Result<(), BoxError>
-where
-    S: Service<peer::HandshakeRequest, Response = peer::Client, Error = BoxError> + Clone,
-    S::Future: Send + 'static,
-{
-    info!("Trying to open Zcash protocol endpoint at {}...", addr);
-    let listener_result = TcpListener::bind(addr).await;
+/// If the listener is configured to use an automatically chosen port (port `0`),
+/// then the returned address will contain the actual port.
+///
+/// # Panics
+///
+/// If opening the listener fails.
+#[instrument(skip(config), fields(addr = ?config.listen_addr))]
+async fn open_listener(config: &Config) -> (TcpListener, SocketAddr) {
+    // Warn if we're configured using the wrong network port.
+    use Network::*;
+    let wrong_net = match config.network {
+        Mainnet => Testnet,
+        Testnet => Mainnet,
+    };
+    if config.listen_addr.port() == wrong_net.default_port() {
+        warn!(
+            "We are configured with port {} for {:?}, but that port is the default port for {:?}",
+            config.listen_addr.port(),
+            config.network,
+            wrong_net
+        );
+    }
+
+    info!(
+        "Trying to open Zcash protocol endpoint at {}...",
+        config.listen_addr
+    );
+    let listener_result = TcpListener::bind(config.listen_addr).await;
 
     let listener = match listener_result {
         Ok(l) => l,
         Err(e) => panic!(
             "Opening Zcash network protocol listener {:?} failed: {:?}. \
              Hint: Check if another zebrad or zcashd process is running. \
              Try changing the network listen_addr in the Zebra config.",
-            addr, e,
+            config.listen_addr, e,
         ),
     };
 
-    let local_addr = listener.local_addr()?;
+    let local_addr = listener
+        .local_addr()
+        .expect("unexpected missing local addr for open listener");
     info!("Opened Zcash protocol endpoint at {}", local_addr);
+
+    (listener, local_addr)
+}
+
+/// Listens for peer connections on `addr`, then sets up each connection as a
+/// Zcash peer.
+///
+/// Uses `handshaker` to perform a Zcash network protocol handshake, and sends
+/// the `Client` result over `tx`.
```

Comment:
Not sure if `Client` should be [`PeerChange`]? If not, maybe it's still worth adding an intra-doc link?

```suggestion
/// the [`Client`][peer::Client] result over `tx`.
```

Reply:
`PeerChange` is a local type alias:
```rust
type PeerChange = Result<Change<SocketAddr, peer::Client>, BoxError>;
```